docs/metadata.md

```diff
@@ -63,33 +85,41 @@ See [the schema spec](https://github.com/raintank/schema/blob/master/metric.go#L
-* lookup (1) by orgid (2) + target spec, where target spec is:
-  - a graphite key
-  - a graphite pattern that has wildcards (`*`), one of multiple options `{foo,bar}`, character lists `[abc]` and ranges `[a-z0-9]`.
-  - in the future we will want to extend these with tag constraints (e.g. must have given key, key must have given value, or value for key must match a pattern similar to above pattern)
-* prefix search of prefix pattern and orgid (2). this is a special case of a pattern search, but common for autocomplete/suggest with short prefix patterns.
-* listing (e.g. graphite's metrics.json but possibly in more detail for other tools) based on org-id (2).
-* in the future we may also do queries on tags such as:
-  - list all tags
-  - list all tags for a given series pattern
+An index plugin just needs to implement the github.com/raintank/metrictank/idx/MetricIndex interface.
 
+```
+type Node struct {
+	Path string
+	Leaf bool
+	Defs []schema.MetricDefinition
+}
+type MetricIndex interface {
+	Init(met.Backend) error
+	Stop()
+	Add(*schema.MetricData)
+	Get(string) (schema.MetricDefinition, error)
+	List(int) []schema.MetricDefinition
+	Find(int, string) []Node
+	Delete(int, string)
+}
 
-### Notes
+```
 
-(1) lookup: What do we need to lookup? For now we mainly want/only need interval (for alignRequests), mtype (to figure out the consolidation) and name (for listings),
-but ideally we can lookup the entire definition.  E.g. in the future we may end up determining rollup schema based on org and/or misc tags.  
-(2) org-id: we need to return metrics corresponding to a given org, as well as metrics from org -1, since those are publically visible to everyone.
+* Init(met.Backend): This is the initialization step performed at startup.  This method should block until the index is ready to handle searches.
+* Stop(): This will be called when metrictank is shutting down.
+* Add(*schema.MetricData):  Every metric received will result in a call to this method to ensure the metric has been added to the index.
+* Get(string) (schema.MetricDefinition, error):  This method should return the MetricDefintion with the passed Id.
+* List(int) []schema.MetricDefinition: This method should return all MetricDefinitions for the passed OrgId.  If the passed OrgId is "-1", then all metricDefinitions across all organisations should be returned.
+* Find(int, string) []Node: This method provides searches.  The method is passed an OrgId and a query pattern. This pattern should be handled in the same way Graphite would. https://graphite.readthedocs.io/en/latest/render_api.html#paths-and-wildcards.  Searches should return all nodes that match for the given OrgId and OrgId -1.
+* Delete(int, string): This method is used for deleting items from the index. The method is passed an OrgId and a query pattern.  If the pattern matches a branch node, then all leaf nodes on that branch should also be deleted. So if the pattern is "*", all items in the index should be deleted.
 
```

Comment:
let's move this documentation into the actual interface definition. that way we can also easily have it in the code, in godoc, as well as in the docs (by just updating the docs if we update the code)